Add unit tests to test DDP with TensorPipe GPU support.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43940 Add unit tests to test DDP with TensorPipe GPU support.**

DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

pritamdamania committed:
Add unit tests to test DDP with TensorPipe GPU support.

DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add unit tests to test DDP with TensorPipe GPU support."


DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
+        # Run distributed model.
+        with dist_autograd.context() as context_id:
+            res = rpc.rpc_sync(dst, torch.add, args=(t, t))
+            rpc.api._wait_all_workers()
```
Why do we need to wait all workers here? Is it due to the deadlock issue mentioned offline? Should we file an issue describing how the deadlock can happen?

rohan-varma reviewed (approved):
LGTM

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
+
+        # Run distributed model.
+        with dist_autograd.context() as context_id:
+            res = rpc.rpc_sync(dst, torch.add, args=(t, t))
```
Currently, it looks like we just verify remote gradients for an add() and everything else in DDP is local, but ideally the test should cover calling DDP(model) where `model` has both remote and local layers?

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(4)
+    def test_with_ddp(self):
```
I'm not familiar with these tests: what is the reason to add this to the `rpc_test` file rather than to the `ddp_under_dist_autograd_test` file?

lw reviewed (commented):
Should we perhaps wait to land this until we implement the intended stream behavior in the TensorPipe agent (see #44084)?

pritamdamania committed:
Update on "Add unit tests to test DDP with TensorPipe GPU support."


DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

[ghstack-poisoned]

pritamdamania87 commented:
> Should we perhaps wait to land this until we implement the intended stream behavior in the TensorPipe agent (see #44084)?

I think this makes sense since when I enhanced the current unit test based on Rohan's suggestions, we run into a separate synchronization issue since an RPC and allreduce run on the same device and stream concurrently causing a similar as mentioned before deadlock.

pritamdamania committed:
Update on "Add unit tests to test DDP with TensorPipe GPU support."


DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

[ghstack-poisoned]
